### ISSUE TITLE
Should pos-resource support blank nodes?

### ISSUE THREAD

**Issue by josephguillaume** [pod-os/PodOS#35]
There might be an architectural decision to make here. I'll describe my use case and current blockers.

I use pos-resource to make a resource available to descendants. This includes creating pos-resource elements and setting their uri programmatically (usually with lazy=true)

When the resource is a blank node, pos-resource errors with e.g. NamedNode IRI "_g_L9C317" must be absolute

I would need to confirm, but it appears the error is thrown by os.store.get when checking editable, which in turn assumes that the Uri is a named node: https://github.com/linkeddata/rdflib.js/blob/cd9a5960ef01bf9350c55a349db4410bc029a796/src/update-manager.ts#L154

There are other places that the PodOs code currently assumes a Uri is a named node/sym, e.g. https://github.com/pod-os/PodOS/blob/e4904287890fa3c03a37fb485d0b78ddd893cc34/core/src/thing/Thing.ts#L99

Current workaround is to rewrite data to avoid using blank nodes.

Issues to address:
- Should it be possible to specify a blank node as URI for pos-resource (with lazy=true)?
- If yes, how should editable status be determined for blank nodes? E.g. how is this usually handled by rdflib.js?
- If yes, what other changes/tests are required?
- If no, what is the preferred mechanism to access the relations of a blank node within PodOS?

**Comment by angelo-v** [comment]
You are right, PodOS mostly assumes named nodes currently and I did not spent much thoughts on blank nodes(yet). I am trying to avoid them as described in your workaround.

Some things come to my mind, but the whole blank node situation must be carefully thought through:

- a blank node is also a resource just without a name (URI), so pos-resource could apply
- using `uri` attribute to refer to a blank node is definitely not right, we either need to rename it or add another attribute to handle blank node
- how would one refer to blank nodes?
- pos-resource does 2 things: fetch and set the context for siblings. Fetching does not apply to blank nodes
- we also need to consider collections, it is another but maybe similar topic

**Comment by josephguillaume** [comment]
My current use case actually relates to collections - my list component creates pos-resource elements and sets their uri. Calling lazy=true suppresses fetching but allows descendants to receive the resource.

If it's a blank node, the uri starts with an underscore, e.g. _g_L9C317

Options for renaming would be resource or about. consistent with https://www.w3.org/TR/rdfa-lite/#resource or https://www.w3.org/TR/rdfa-primer/#alternative-for-setting-the-context-about
PodOS has the potential for its DOM to become valid RDFa.

Just checking now, it turns out RDFa supports explicitly defining blank nodes, which would work well for our case here.
https://www.w3.org/TR/rdfa-core/#s_blankNodes

**Comment by jg10-mastodon-social** [comment]
Documenting another use case: ActivityPub uses `https://www.w3.org/ns/activitystreams#attachment` and in the context of json-ld/activity+json, this links to a blank node with further properties, e.g.
```turtle
_:b4 <http://joinmastodon.org/ns#blurhash> "U%JRdVj[fQof~qoft7j[xufQofj[%MfQayay" .
_:b4 <http://joinmastodon.org/ns#focalPoint> _:b6 .
_:b4 <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <https://www.w3.org/ns/activitystreams#Document> .
_:b4 <https://www.w3.org/ns/activitystreams#height> "1001"^^<http://www.w3.org/2001/XMLSchema#nonNegativeInteger> .
_:b4 <https://www.w3.org/ns/activitystreams#mediaType> "image/png" .
_:b4 <https://www.w3.org/ns/activitystreams#url> <https://media.mastodon.com.br/media_attachments/files/115/698/696/714/587/951/original/95ddb90b6f2a3f21.png> .
_:b4 <https://www.w3.org/ns/activitystreams#width> "635"^^<http://www.w3.org/2001/XMLSchema#nonNegativeInteger> .
```

**Comment by josephguillaume** [comment]
It turns out that the error "NamedNode IRI "_g_L9C317" must be absolute" is actually because the string version of the blank node is "_:_g_L9C317", which doesn't cause an error.

`pos-resource` therefore does support blank nodes out of the box and it appears there was a bug in my list component.

I won't close the issue yet because other parts of the discussion do still seem to apply.
